### PR TITLE
[#463] Add monitor for processing cross links 

### PR DESCRIPTION
**Pull request by janet-harmony**
[watchdog] Add monitor for processing cross links
[watchdog] Fix shard id for committee information
[watchdog] Fix used-seat count for status page
[refactor] One function for getting format for RPC requests
[refactor] Move PagerDuty notification messages into a separate file

### CODE REVIEW

File: monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+// Only need to query leader on Shard 0
+func (m *monitor) crossLinkMonitor(interval uint64, poolSize int, pdServiceKey, chain string, shardMap map[string]int) {
+  crossLinkRequestFields := getRPCRequest(LastCrossLinkRPC)
+  nodeRequestFields := getRPCRequest(NodeMetadataRPC)
+
+  jobs := make(chan work, len(shardMap))
+  replyChannels := make(map[string](chan reply))
+  syncGroups := make(map[string]*sync.WaitGroup)
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go]: why are the sync groups pointers?

File: monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      }
+		}
+		syncGroups[NodeMetadataRPC].Wait()
+		close(replyChannels[NodeMetadataRPC])
+
+		leader := []string{}
+		for d := range replyChannels[NodeMetadataRPC] {
+			if d.oops == nil {
+				oneReport := r{}
+				json.Unmarshal(d.rpcResult, &oneReport)
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go]: handle this error

File: monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    // Request from all potential leaders
+    for _, l := range leader {
+      crossLinkRequestFields["id"] = strconv.Itoa(queryID)
+      requestBody, _ := json.Marshal(crossLinkRequestFields)
+      jobs <- work{l, LastCrossLinkRPC, requestBody}
+      queryID++
+      syncGroups[LastCrossLinkRPC].Add(1)
+    }
+    syncGroups[LastCrossLinkRPC].Wait()
+		close(replyChannels[LastCrossLinkRPC])
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go]: please come around and use gofmt as a plugin in your editor. have it run on file save

File: monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go [hunk 4]
```diff
@@ -931,7 +890,7 @@ func (m *monitor) statusSnapshot() statusReport {
 
 	usedSeats := 0
 	for _, info := range m.SuperCommittee.CurrentCommittee.Deciders {
-		usedSeats =+ linq.From(info.Committee).CountWith(func (v interface{}) bool {
+		usedSeats += linq.From(info.Committee).CountWith(func (v interface{}) bool {
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go]: okay til. what's the diff?

File: monitoring/harmony-monitor/blockchain-watchdog/rpc.go [hunk 5]
```diff
@@ -4,12 +4,13 @@ package main
...
 	BlockHeaderRPC    = "hmy_latestHeader"
-	PendingCxRPC      = "hmy_getPendingCxReceipts"
+	PendingCXRPC      = "hmy_getPendingCXReceipts"
 	SuperCommitteeRPC = "hmy_getSuperCommittees"
+	LastCrossLinkRPC  = "hmy_getLastCrossLinks"
 	JSONVersion       = "2.0"
 )
 
-// NodeMetadataReply
+// Node Metadata RPC Reply
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/rpc.go]: install golint as a plugin for editor, this is actually golang convention you're going against.

File: monitoring/harmony-monitor/blockchain-watchdog/rpc.go [hunk 6]
```diff
@@ -64,20 +82,41 @@ type SuperCommitteeReply struct {
...
+	VotingPower     string           `json:"voting-power-%"`
+	EffectiveStake  string           `json:"effective-stake,omitempty"`
+}
+
+// Last Cross Links RPC Reply
+type LastCrossLinkReply struct {
+	CrossLinks      []CrossLink      `json:"-"`
+}
+
+type CrossLink struct {
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/rpc.go]: Let's build watchdog statically against the main repo so we can get rid of these redundant data strucutres and just import them directly from main repo

File: monitoring/harmony-monitor/blockchain-watchdog/rpc.go [hunk 7]
```diff
@@ -64,20 +82,41 @@ type SuperCommitteeReply struct {
...
+	IsHarmonyNode   bool             `json:"is-harmony-slot"`
+	BLSKey          string           `json:"bls-public-key"`
+	RawPercent      string           `json:"voting-power-unnormalized,omitempty"`
+	VotingPower     string           `json:"voting-power-%"`
+	EffectiveStake  string           `json:"effective-stake,omitempty"`
+}
+
+// Last Cross Links RPC Reply
+type LastCrossLinkReply struct {
+	CrossLinks      []CrossLink      `json:"-"`
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/rpc.go]: is this even here if you're ignoring the json in result?

File: monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go (hunk 4 above)
Comment (janet-harmony) [reply monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go]: First one is the same as `usedSeats = +(linq...`
Seconds actually adds the `linq` value to `usedSeats`

File: monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go (hunk 2 above)
Comment (janet-harmony) [reply monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go]: No-op if node doesn't reply.

File: monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go (hunk 3 above)
Comment (janet-harmony) [reply monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go]: Forgot to run gofmt on the file.

**Comment by fxfactorial** [comment]
Wait the super committee rpc reply looks like it still using a %, I thought I changed that . Double check the super committee rpc

**Comment by janet-harmony** [comment]
```
"committee-members": [
            {
              "bls-public-key": "a6e2ae873327f2c5fc11c94bdb874ba82344a33cad8d0081600063278bd2822024d905e6338a2b8a8f84eef15f20fa14",
              "earning-account": "one10jvjrtwpz2sux2ngktg3kq7m3sdz5p5au5l8c8",
              "effective-stake": "31050000000000000000000.000000000000000000",
              "is-harmony-slot": false,
              "voting-power-%": "0.053419354838709677",
              "voting-power-unnormalized": "0.166935483870967742"
            },
```

**Pull request merged**